### PR TITLE
[Xamarin.Android.Build.Tasks] Prevent `$(AndroidUseLatestPlatformSdk)` from using Unstable APIs

### LINKED ISSUE #1221: Prevent `$(AndroidUseLatestPlatformSdk)` from using Unstable APIs
Currently if a user sets `$(AndroidUseLatestPlatformSdk)` the system will pickup the very latest binding target framework, regardless of if it is stable or not. 

What we should be doing is using the latest *Stable* API rather than just grabbing the max. 
That said we still want the user to be able to use the unstable API if they really want to. So regardless of `$(AndroidUseLatestPlatformSdk)` is set to true, if the user has set the `$(TargetFrameworkVersion)` to the value of an unstable API we should allow the user to use it. 

So the new logic is as follows

1. `$(AndroidUseLatestPlatformSdk)` = true, use the latest stable API
2. `$(TargetFrameworkVersion)` = and unstable version and `$(AndroidUseLatestPlatformSdk)` = false, use the `$(TargetFrameworkVersion)`.
3.`$(TargetFrameworkVersion)` = and unstable version and `$(AndroidUseLatestPlatformSdk)` = true, use the `$(TargetFrameworkVersion)`

### PR DESCRIPTION
**Pull request by dellis1972**
Fixes #1221

Currently when a user sets `$(AndroidUseLatestPlatformSdk)` it will
automatically use the Max ApiLevel. This is not always a stable
ApiLevel. It is highly possible that it will be unstable.
Users will generally want to keep to the Max Stable in this senario.

This commit alters the code so that we use the latest Max Stable
ApiLevel when using `$(AndroidUseLatestPlatformSdk)`.
However we should still allow users to use the unstable ones if
they really want to. This can be achieved in a number of ways

1. Turn off `$(AndroidUseLatestPlatformSdk)` and set the
`$(TargetFrameworkVersion)` directly.

2. Leave `$(AndroidUseLatestPlatformSdk)` on and set the
`$(TargetFrameworkVersion)` directly.

So effectviely, just manually target the unstable Api. But it
is something the user will need to think about and do manaully.
By default we will only use to the latest stable ApiLevel.

### CODE REVIEW

**Comment by dellis1972** [comment]
build

**Comment by jonpryor** [comment]
The [**macOS+xbuild PR Build**](https://jenkins.mono-project.com/job/xamarin-android-pr-builder/2414/) is [failing](https://jenkins.mono-project.com/job/xamarin-android-pr-builder/2414/testReport/junit/Xamarin.Android.Build.Tests/BuildTest/ValidateUseLatestAndroid___Debug/):

```
MESSAGE: Build should have succeeded
  Expected: True
  But was:  False
```

**Comment by dellis1972** [comment]
build

File: src/Xamarin.Android.Build.Tasks/Tasks/ResolveSdksTask.cs [hunk 1]
```diff
@@ -423,13 +423,21 @@ bool ValidateApiLevels ()
...
 				if (int.TryParse (AndroidApiLevel, out maxInstalled) && int.TryParse (SupportedApiLevel, out maxSupported) && maxInstalled > maxSupported) {
 					Log.LogDebugMessage ($"API Level {AndroidApiLevel} is greater than the maximum supported API level of {SupportedApiLevel}. " +
 						"Support for this API will be added in a future release.");
 					AndroidApiLevel = SupportedApiLevel;
 				}
+				if (!string.IsNullOrWhiteSpace (TargetFrameworkVersion)) {
+					var userSelected = MonoAndroidHelper.SupportedVersions.GetApiLevelFromFrameworkVersion (TargetFrameworkVersion);
+					// overwrite using user version only if it is 
+					// above the maxStableApi and a valid apiLevel.
+					if (userSelected != null && userSelected > maxSupported && userSelected <= maxInstalled) {
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Tasks/ResolveSdksTask.cs]: None of this make sense. :-(

The use case is this: "soon", Google will presumably release API-P in alpha form, for release as API-28 this fall.

When we provide a binding, it will *probably* be `$(TargetFrameworkVersion)`=v8.99.

Assuming the user has `v8.99` installed -- *how* that's done is "elsewhere" -- the user should be able to opt in to using it by setting `$(TargetFrameworkVersion)`=v8.99.

Thus, `!string.IsNullOrWhiteSpace (TargetFrameworkVersion)` (line 433) will be `true`.

`userSelected` (line 434), meanwhile, will be `null`, as *there is no API level for unstable APIs*.  Relatedly, `MonoAndroidHelper.SupportedVersions.GetIdFromFrameworkVersion("v8.99")` will return `"P"`.

Similarly, `maxInstalled` comes from `int.TryParse()` comes from `GetMaxInstalledApiLevel()`, which only looks at values which can be parsed into an `int`. As such, *API-P will not be present or used*, meaning -- for our current exercise -- `AndroidApiLevel` is ~useless. Consequently, `maxInstalled` is likewise useless.

In short, I don't see how this will do what we want it to do. :-(

File: src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/BuildTest.cs [hunk 2]
```diff
@@ -2228,6 +2228,56 @@ public void IfAndroidJarDoesNotExistThrowXA5207 ()
...
 
+		[Test]
+		public void ValidateUseLatestAndroid ()
+		{
+			var path = Path.Combine ("temp", TestName);
+			var referencesPath = CreateFauxReferencesDirectory (Path.Combine (path, "xbuild-frameworks"), new ApiInfo [] {
+				new ApiInfo () { Id = 23, Level = 23, Name = "Marshmallow", FrameworkVersion = "v6.0", Stable = true },
+				new ApiInfo () { Id = 26, Level = 26, Name = "Oreo", FrameworkVersion = "v8.0", Stable = true },
+				new ApiInfo () { Id = 27, Level = 27, Name = "Oreo", FrameworkVersion = "v8.1", Stable = false },
+			});
```
Comment (jonpryor) [inline src/Xamarin.Android.Build.Tasks/Tests/Xamarin.Android.Build.Tests/BuildTest.cs]: For the test case, we need to add:

```csharp
new ApiInfo () { Id = "P", Level = 28, Name = "P", FrameworkVersion="v8.99", Stable = false },
```

(This is why `AndroidApiInfo.xml` differentiates `Id` from `Level`. They're the same for all *stable* API levels, but for preview API levels *they are not*.)

File: tools/scripts/xabuild [hunk 3]
```diff
@@ -133,7 +133,11 @@ case "$MSBUILD" in
 		XABUILD_FLAGS+=(/p:TargetFrameworkRootPath="$TARGETS_DIR-frameworks/")
 		;;
 	*xbuild*)
-		export XBUILD_FRAMEWORK_FOLDERS_PATH="$TARGETS_DIR-frameworks"
+		if [ -z "$XBUILD_FRAMEWORK_FOLDERS_PATH" ]; then
+			export XBUILD_FRAMEWORK_FOLDERS_PATH="$TARGETS_DIR-frameworks"
+		else 
+			export XBUILD_FRAMEWORK_FOLDERS_PATH="$XBUILD_FRAMEWORK_FOLDERS_PATH"
```
Comment (jonpryor) [inline tools/scripts/xabuild]: Why does the environment have `$XBUILD_FRAMEWORK_FOLDERS_PATH` already exported, and why should we be using it *in preference to the computed value*?

Comment (dellis1972) [reply tools/scripts/xabuild]: We need to be able to override it to the Faux location as part of the unit tests. If we don't it will pick up the `bin\Debug` path which we don't want. Is there a better way of doing it?

**Comment by dellis1972** [comment]
if we merge https://github.com/xamarin/xamarin-android/pull/1321 I can add actual tests for this logic :)

File: tools/scripts/xabuild (hunk 3 above)
Comment (jonpryor) [reply tools/scripts/xabuild]: No, I can't think of a better way, though this will only work with `xbuild`, not `msbuild`.

Perhaps a "better" (or just different?) way would be to define a *new* environment variable, e.g. `$__MSBUILD_FRAMEWORK_FOLDERS_PATH_OVERRIDE`, and use that if it's set. We'd also want to update the `msbuild`-based codepath to follow suit, perhaps via [setting `$(TargetFrameworkRootPath)`](https://github.com/xamarin/xamarin-android/blob/49ecddaea61d789486d5d7b9724d842d495f1257/tools/scripts/xabuild#L133)?

One problem with the current `xabuild` script, though: the `msbuild`-specific logic on line 133 happens *after* we [`exec xabuild.exe` on line 86](https://github.com/xamarin/xamarin-android/blob/49ecddaea61d789486d5d7b9724d842d495f1257/tools/scripts/xabuild#L86), meaning line 133/etc. is superfluous. This should be corrected.

Comment (dellis1972) [reply tools/scripts/xabuild]: msbuild will use the `TargetFrameworkRootPath` as a parameter so msbuild support works. xbuild ignores or doesn't use `TargetFrameworkRootPath` so thats why I needed to use the envvar.

**Comment by dellis1972** [comment]
@jonpryor  Given the following bindings are available  can you confirm the correct behaviour (assuming the required java platforms are installed).

`v8.0` => API 26
`v8.1` => API 27
`v8.99` => API 28 (P) 

| UseLatestPlatformSdk | TargetFrameworkVersion | Result |
| :----------------------: |:-------------------------:|:--------:|
| True | v8.99 | v8.99 |
| True | v8.1 | v8.1 |
| True | v8.0 | v8.1 |
| True | v6.0 | v8.1 |
| False | v8.99 | v8.99 |
| False | v8.1 | v8.1 |
| False | v8.0 | v8.0 |
| False | v6.0 | Error |

**Comment by jonpryor** [comment]
@dellis1972: that table *mostly* looks correct. However, this line seems weird:

| UseLatestPlatformSdk | TargetFrameworkVersion | Result |
| :----------------------: |:-------------------------:|:--------:|
| False | null | v8.1 |

If `$(TargetFrameworkVersion)` isn't set, and `$(UseLatestPlatformSdk)` is False, shouldn't *this* be an XA0001 error, as `null` is *not* a valid `$(TargetFrameworkVersion)` value?
